File: android/src/main/java/com/adobe/marketing/mobile/flutter/FlutterACPPlacesPlugin.java
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
package com.adobe.marketing.mobile.flutter;

import android.location.Location;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.adobe.marketing.mobile.AdobeCallback;
import com.adobe.marketing.mobile.Places;
import com.adobe.marketing.mobile.PlacesAuthorizationStatus;
import com.adobe.marketing.mobile.PlacesPOI;
import com.google.android.gms.location.Geofence;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * FlutterACPPlacesPlugin
 */
public class FlutterACPPlacesPlugin implements FlutterPlugin, MethodCallHandler {

    final static String METHOD_PLACES_CLEAR = "clear";
    final static String METHOD_PLACES_EXTENSION_VERSION = "extensionVersion";
    final static String METHOD_PLACES_GET_CURRENT_POINTS_OF_INTEREST = "getCurrentPointsOfInterest";
    final static String METHOD_PLACES_GET_LAST_KNOWN_LOCATION = "getLastKnownLocation";
    final static String METHOD_PLACES_GET_NEARBY_POINTS_OF_INTEREST = "getNearbyPointsOfInterest";
    final static String METHOD_PLACES_PROCESS_GEOFENCE = "processGeofence";
    final static String METHOD_PLACES_SET_AUTHORIZATION_STATUS = "setAuthorizationStatus";

    final static String LOG_TAG = "ACPPlaces_Flutter";

    final static String POI = "POI";
    final static String LATITUDE = "latitude";
    final static String LONGITUDE = "longitude";
    final static String IDENTIFIER = "identifier";
    final static String LIMIT = "Limit";
    final static String LOCATION = "Location";
    final static String GEOFENCE = "Geofence";
    final static String TRANSITION_TYPE = "TransitionType";
    final static String RADIUS = "radius";
    final static String EXPIRATION_DURATION = "expirationDuration";
    final static String REQUEST_ID = "requestId";
    private MethodChannel channel;

    @Override
    public void onAttachedToEngine(@NonNull final FlutterPluginBinding binding) {
        channel = new MethodChannel(binding.getBinaryMessenger(), "flutter_acpplaces");
        channel.setMethodCallHandler(new FlutterACPPlacesPlugin());
    }

    @Override
    public void onDetachedFromEngine(@NonNull final FlutterPluginBinding binding) {
        if (channel != null) {
            channel.setMethodCallHandler(null);
        }
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        if (METHOD_PLACES_EXTENSION_VERSION.equals(call.method)) {
            result.success(Places.extensionVersion());
        } else if (METHOD_PLACES_CLEAR.equals((call.method))) {
            Places.clear();
            result.success(null);
        } else if (METHOD_PLACES_GET_CURRENT_POINTS_OF_INTEREST.equals((call.method))) {
            getCurrentPointsOfInterest(result);
        } else if (METHOD_PLACES_GET_LAST_KNOWN_LOCATION.equals((call.method))) {
            getLastKnownLocation(result);
        } else if (METHOD_PLACES_GET_NEARBY_POINTS_OF_INTEREST.equals((call.method))) {
            getNearbyPointsOfInterest(result, call.arguments);
        } else if (METHOD_PLACES_PROCESS_GEOFENCE.equals((call.method))) {
            processGeofence(call.arguments);
            result.success(null);
        } else if (METHOD_PLACES_SET_AUTHORIZATION_STATUS.equals(call.method)) {
            setAuthorizationStatus(call.arguments);
            result.success(null);
        } else {
            result.notImplemented();
        }
    }

    private void getCurrentPointsOfInterest(final Result result) {
        Places.getCurrentPointsOfInterest(new AdobeCallback<List<PlacesPOI>>() {
            @Override
            public void call(final List<PlacesPOI> pois) {
                runOnUIThread(new Runnable() {
                    @Override
                    public void run() {
                        result.success(generatePOIString(pois));
                    }
                });
            }
        });
    }

    private void getLastKnownLocation(final Result result) {
        Places.getLastKnownLocation(new AdobeCallback<Location>() {
            @Override
            public void call(final Location location) {
                runOnUIThread(new Runnable() {
                    @Override
                    public void run() {
                        if (location != null) {
                            JSONObject json = new JSONObject();
                            try {
                                json.put(LATITUDE, location.getLatitude());
                                json.put(LONGITUDE, location.getLongitude());
                            } catch (JSONException e) {
                                Log.e(LOG_TAG, "Error putting data into JSON: " + e.getLocalizedMessage());
                            }
                            result.success(json.toString());
                        }
                    }
                });
            }
        });
    }

    private void getNearbyPointsOfInterest(final Result result, final Object arguments) {
        if (!(arguments instanceof Map)) {
            Log.e(LOG_TAG, "Get Nearby Points of Interest failed because arguments were invalid");
            return;
        }
        Map<String, Object> params = (Map<String, Object>) arguments;
        if (!(params.get(LOCATION) instanceof Map) || !(params.get(LIMIT) instanceof Integer)) {
            Log.e(LOG_TAG, "Get Nearby Points of Interest failed because arguments were invalid");
            return;
        }

        final Location location = new Location("default provider");
        final Map<String, Object> locationMap = (Map<String, Object>) params.get(LOCATION);
        location.setLatitude((Double) locationMap.get(LATITUDE));
        location.setLongitude((Double) locationMap.get(LONGITUDE));
        final int limit = (int) params.get(LIMIT);

        Places.getNearbyPointsOfInterest(location, limit, new AdobeCallback<List<PlacesPOI>>() {
            @Override
            public void call(final List<PlacesPOI> pois) {
                runOnUIThread(new Runnable() {
                    @Override
                    public void run() {
                        result.success(generatePOIString(pois));
                    }
                });
            }
        });
    }

    private void processGeofence(final Object arguments) {
        if (!(arguments instanceof Map)) {
            Log.e(LOG_TAG, "Process Geofence failed because arguments were invalid");
            return;
        }
        Map<String, Object> params = (Map<String, Object>) arguments;
        if (!(params.get(GEOFENCE) instanceof Map) || !(params.get(TRANSITION_TYPE) instanceof Integer)) {
            Log.e(LOG_TAG, "Process Geofence failed because arguments were invalid");
            return;
        }

        final Map<String, Object> geofenceMap = (Map<String, Object>) params.get(GEOFENCE);
        final Integer radius = (Integer) geofenceMap.get(RADIUS);
        final Integer expirationDuration = (Integer) geofenceMap.get(EXPIRATION_DURATION);
        final int transitionType = (int) params.get(TRANSITION_TYPE);
        final Geofence geofence = new Geofence.Builder()
                .setCircularRegion((Double) geofenceMap.get(LATITUDE), (Double) geofenceMap.get(LONGITUDE), radius.floatValue())
                .setExpirationDuration(expirationDuration.longValue())
                .setTransitionTypes(transitionType)
                .setRequestId((String) geofenceMap.get(REQUEST_ID))
                .build();
        Places.processGeofence(geofence, transitionType);
    }

    private void setAuthorizationStatus(final Object arguments) {
        if (!(arguments instanceof Integer)) {
            Log.e(LOG_TAG, "Set Authorization Status failed because arguments were invalid");
            return;
        }

        Places.setAuthorizationStatus(getAuthorizationStatus((Integer) arguments));
    }

    private void runOnUIThread(Runnable runnable) {
        new Handler(Looper.getMainLooper()).post(runnable);
    }

    private String generatePOIString(final List<PlacesPOI> pois) {
        JSONArray jsonArray = new JSONArray();
        JSONObject json;
        if (!pois.isEmpty()) {
            for (int index = 0; index < pois.size(); index++) {
                try {
                    PlacesPOI poi = pois.get(index);
                    json = new JSONObject();
                    json.put(POI, poi.getName());
                    json.put(LATITUDE, poi.getLatitude());
                    json.put(LONGITUDE, poi.getLongitude());
                    json.put(IDENTIFIER, poi.getIdentifier());
                    jsonArray.put(index, json);
                } catch (JSONException e) {
                    Log.d(LOG_TAG, "Error putting data into JSON: " + e.getLocalizedMessage());
                }
            }
        }
        return jsonArray.toString();
    }

    private PlacesAuthorizationStatus getAuthorizationStatus(final int status) {
        if (status == 0) {
            return PlacesAuthorizationStatus.DENIED;
        } else if (status == 1) {
            return PlacesAuthorizationStatus.ALWAYS;
        } else if (status == 2) {
            return PlacesAuthorizationStatus.UNKNOWN;
        } else if (status == 3) {
            return PlacesAuthorizationStatus.RESTRICTED;
        } else if (status == 4) {
            return PlacesAuthorizationStatus.WHEN_IN_USE;
        } else {
            return null;
        }
    }
}

File: ios/Classes/FlutterACPPlacesPlugin.m
/*
Copyright 2020 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/
#import "FlutterACPPlacesPlugin.h"
#import "ACPPlaces.h"

static NSString* const METHOD_PLACES_CLEAR = @"clear";
static NSString* const METHOD_PLACES_EXTENSION_VERSION = @"extensionVersion";
static NSString* const METHOD_PLACES_GET_CURRENT_POINTS_OF_INTEREST = @"getCurrentPointsOfInterest";
static NSString* const METHOD_PLACES_GET_LAST_KNOWN_LOCATION = @"getLastKnownLocation";
static NSString* const METHOD_PLACES_GET_NEARBY_POINTS_OF_INTEREST = @"getNearbyPointsOfInterest";
static NSString* const METHOD_PLACES_PROCESS_GEOFENCE = @"processGeofence";
static NSString* const METHOD_PLACES_SET_AUTHORIZATION_STATUS = @"setAuthorizationStatus";
static NSString* const POI = @"POI";
static NSString* const IDENTIFIER = @"Identifier";
static NSString* const EMPTY_ARRAY_STRING = @"[]";
static NSString* const LATITUDE = @"latitude";
static NSString* const LONGITUDE = @"longitude";
static NSString* const CIRCULAR_REGION = @"circularRegion";
static NSString* const RADIUS = @"radius";
static NSString* const REQUEST_ID = @"requestId";
static NSString* const TRANSITION_TYPE = @"TransitionType";
static NSString* const LIMIT = @"Limit";
static NSString* const LOCATION = @"Location";
static NSString* const GEOFENCE = @"Geofence";

@implementation FlutterACPPlacesPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"flutter_acpplaces"
            binaryMessenger:[registrar messenger]];
  FlutterACPPlacesPlugin* instance = [[FlutterACPPlacesPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([METHOD_PLACES_EXTENSION_VERSION isEqualToString:call.method]) {
        result([ACPPlaces extensionVersion]);
    } else if ([METHOD_PLACES_CLEAR isEqualToString:call.method]) {
        [ACPPlaces clear];
        result(nil);
    } else if ([METHOD_PLACES_GET_CURRENT_POINTS_OF_INTEREST isEqualToString:call.method]) {
        [self getCurrentPointsOfInterest:call result:result];
    } else if ([METHOD_PLACES_GET_LAST_KNOWN_LOCATION isEqualToString:call.method]) {
        [self getLastKnownLocation:call result:result];
    } else if ([METHOD_PLACES_GET_NEARBY_POINTS_OF_INTEREST isEqualToString:call.method]) {
        [self getNearbyPointsOfInterest:call result:result];
    } else if ([METHOD_PLACES_PROCESS_GEOFENCE isEqualToString:call.method]) {
        [self processGeofence:call];
        result(nil);
    } else if ([METHOD_PLACES_SET_AUTHORIZATION_STATUS isEqualToString:call.method]) {
        NSNumber* status = call.arguments;
        [ACPPlaces setAuthorizationStatus:[self convertToCLAuthorizationStatus:[status intValue]]];
        result(nil);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)getCurrentPointsOfInterest:(FlutterMethodCall *)call result:(FlutterResult)result {
    __block NSString* currentPoisString = EMPTY_ARRAY_STRING;
    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
    [ACPPlaces getCurrentPointsOfInterest:^(NSArray<ACPPlacesPoi *> * _Nullable retrievedPois) {
        if(retrievedPois != nil && retrievedPois.count != 0) {
        currentPoisString = [self generatePOIString:retrievedPois];
        dispatch_semaphore_signal(semaphore);
        }
    }];
    dispatch_semaphore_wait(semaphore, dispatch_time(DISPATCH_TIME_NOW, ((int64_t)1 * NSEC_PER_SEC)));
    result(currentPoisString);
}

- (void)getLastKnownLocation:(FlutterMethodCall *)call result:(FlutterResult)result {
    [ACPPlaces getLastKnownLocation:^(CLLocation * _Nullable lastLocation) {
        NSMutableDictionary* tempDict = [[NSMutableDictionary alloc]init];
        [tempDict setValue:[NSNumber numberWithDouble:lastLocation.coordinate.latitude] forKey:LATITUDE];
        [tempDict setValue:[NSNumber numberWithDouble:lastLocation.coordinate.longitude] forKey:LONGITUDE];
        NSData* jsonData = [NSJSONSerialization dataWithJSONObject:tempDict options:0 error:nil];
        result([[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding]);
    }];
}

- (void)getNearbyPointsOfInterest:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary* argumentDict = (NSDictionary *) call.arguments;
    NSDictionary* locationDict = [argumentDict valueForKey:LOCATION];
    CLLocationDegrees latitude = [[locationDict valueForKey:LATITUDE] doubleValue];
    CLLocationDegrees longitude = [[locationDict valueForKey:LONGITUDE] doubleValue];
    CLLocation* currentLocation = [[CLLocation alloc] initWithLatitude:latitude longitude:longitude];
    NSUInteger limit = [[argumentDict valueForKey:LIMIT] integerValue];
    __block NSString* currentPoisString = EMPTY_ARRAY_STRING;
    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
    [ACPPlaces getNearbyPointsOfInterest:currentLocation limit:limit callback:^(NSArray<ACPPlacesPoi *> * _Nullable retrievedPois) {
        currentPoisString = [self generatePOIString:retrievedPois];
        dispatch_semaphore_signal(semaphore);
    }
    errorCallback:^(ACPPlacesRequestError error) {
        result([NSString stringWithFormat:@"Places request error code: %lu", error]);
    }];
    dispatch_semaphore_wait(semaphore, dispatch_time(DISPATCH_TIME_NOW, ((int64_t)1 * NSEC_PER_SEC)));
    result(currentPoisString);
}

- (void)processGeofence:(FlutterMethodCall *)call {
    NSDictionary* argumentDict = (NSDictionary *) call.arguments;
    NSDictionary* geofenceDict = [argumentDict valueForKey:GEOFENCE];
    ACPRegionEventType eventType = [[argumentDict valueForKey:TRANSITION_TYPE] integerValue];
    CLLocationDegrees latitude = [[geofenceDict valueForKey:LATITUDE] doubleValue];
    CLLocationDegrees longitude = [[geofenceDict valueForKey:LONGITUDE] doubleValue];
    CLLocationCoordinate2D center = CLLocationCoordinate2DMake(latitude,longitude);
    NSUInteger radius = [[geofenceDict valueForKey:RADIUS] integerValue];
    NSString* identifier = [geofenceDict valueForKey:REQUEST_ID];
    CLRegion* region = [[CLCircularRegion alloc] initWithCenter:center radius:radius identifier:identifier];
    [ACPPlaces processRegionEvent:region forRegionEventType:eventType];
}

/*
 * Helper functions
 */

- (NSString*) generatePOIString:(NSArray<ACPPlacesPoi *> *) retrievedPois {
    NSMutableArray* retrievedPoisArray = [[NSMutableArray alloc]init];
    if(retrievedPois != nil && retrievedPois.count != 0) {
        for (int index = 0; index < retrievedPois.count; index++) {
            NSMutableDictionary* tempDict = [[NSMutableDictionary alloc]init];
            ACPPlacesPoi* currentPoi = retrievedPois[index];
            [tempDict setValue:currentPoi.name forKey:POI];
            [tempDict setValue:[NSNumber numberWithDouble:currentPoi.latitude] forKey:LATITUDE];
            [tempDict setValue:[NSNumber numberWithDouble:currentPoi.longitude] forKey:LONGITUDE];
            [tempDict setValue:currentPoi.identifier forKey:IDENTIFIER];
            retrievedPoisArray[index] = tempDict;
        }
        NSData* jsonData = [NSJSONSerialization dataWithJSONObject:retrievedPoisArray options:0 error:nil];
        return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    }
    return EMPTY_ARRAY_STRING;
}

- (CLAuthorizationStatus) convertToCLAuthorizationStatus:(int) status {
    switch (status) {
    case 0:
        return kCLAuthorizationStatusDenied;
        break;

    case 1:
        return kCLAuthorizationStatusAuthorizedAlways;
        break;

    case 2:
        return kCLAuthorizationStatusNotDetermined;
        break;

    case 3:
        return kCLAuthorizationStatusRestricted;
        break;

    case 4:
    default:
        return kCLAuthorizationStatusAuthorizedWhenInUse;
        break;
    }
}

@end
